### DISCUSSION TITLE
How To Accumulate Data Across Multiple Events And States?

### DISCUSSION THREAD

**Discussion by rs-georg** [nsk90/kstatemachine#112] [Q&A] [answered]
Imagine a form or wizard that has multiple steps. Each steps collects data, then moves to the next step. In the final step, all data collected in the previous steps should be available so it can be processed/uploaded/...

Is this possible? How could it be done?

For example
- Step 1
Get username and email
`machine.processEvent(UserData(username, email))`
--> transition to Step 2
- Step 2
Get city and street
`machine.processEvent(Location(city, street))`
--> transition to Step 3
- Step 3
Ask for favorite food
`machine.processEvent(FavoriteFood("pizza"))`
--> transition to Step 4
- Step 4
Search for restaurant that serves pizza in the provided city close to the street
--> This means we need to have all the data from previous states in this last state

**Comment by nsk90** [comment, marked as answer]
Hi, sorry looks I missed your question. 

There is no some built-in helper for such task, so the most strait-forward way seems to be using State subclasses for each step.
In State subclasses (steps) your can declare properties to hold all the necessary data. And fill it when the step is complete.
 At the last step you can get all that properties from states (by direct references or using `findState` api).
Alternatively you can store all per-step data in some separate data structure (outside of StateMachine).

**Reply by rs-georg** [reply]
Thanks for you input. This is the approach I ended up using. It's not ideal but will do for now. Thanks again!
